New group folders for dynamic map services

tmcgee opened the pull request:
I'm not sure this addresses all issues or scenarios.
Submitting for review to identify what has not been addressed.

![image](https://user-images.githubusercontent.com/200780/28235393-a2fc1d30-68c1-11e7-9329-ecb241abebea.png)

tmcgee committed:
new group folders for dynamic map services.

tmcgee committed:
Change icon for Grouped layers to folder

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

green3g commented on viewer/js/gis/dijit/LayerControl/controls/_DynamicFolder.js:
```diff
@@ -99,18 +97,129 @@ define([
+        },
+
+        _getSubLayerNodes: function () {
+            var layerIds = this.control.controlOptions.layerIds;
```
It looks like this is a bug, if layerIds is not provided, it breaks on 176.

tmcgee replied:
updated line 161 to:

``` javascript
var layerIds = this.control.controlOptions.layerIds || [];
```

green3g commented on viewer/js/gis/dijit/LayerControl/controls/_DynamicFolder.js:
```diff
@@ -99,18 +97,129 @@ define([
+                if (array.indexOf(this.sublayerInfo.subLayerIds, subLayerID) < 0) {
+                    return false;
+                // is the sublayer included in layer's layerIds (if they are defined)
+                } else if (layerIds.length && array.indexOf(layerIds, subLayerID) < 0) {
```
`layerIds.length` throws an error when `layerIds` is not defined

green3g commented:
I think this is a good addition. I might've changed the folder icons to something a little bit darker, its a little hard to see on some screens I think because it is so light, white on white. Just my two cents. 

That being said, I don't think I can personally utilize this in my app at the moment because of the way my groups are set up. Essentially I start out with something like this, where each group layer is off, but several children are turned on in each group. No layers are displayed until the user turns on one or several groups. The users like to be able to turn on the group and get all of their related layers turned on. 

![image](https://user-images.githubusercontent.com/5471079/28689221-7c31cc5e-72da-11e7-9361-ba6e285a1d92.png)


But I've also included some extra layers in the folder that don't get used regularly but are available for people who need them. Toggling the group now, turns on those extra layers. 

That's not a huge issue, I can move them to a different group or something else, there are a few ways to handle it, I just wanted it to be mentioned for future users to find. 

Which leads me to my second issue:
 I'm not sure why at the moment but when I turn on the folder, like Documentation, it turns on all the other group layers, which renders about 30 layers. I wouldn't expect this to happen and it looks to me like a bug. I'll need to look into it a bit more. I can pm you with a link to the service to test on.

tmcgee committed:
ensure that layerIds variable is assigned a value..

tmcgee committed:
re-initialize the groupedLayerInfos once the grouped layers have been added.
fixes #754.

tmcgee committed:
re-initialize the groupedLayerInfos once the grouped layers have been added.
fixes #747

tmcgee committed:
Merge branch 'feature/new-group-folders-for-dynamic-map-services' of github.com:cmv/cmv-app into feature/new-group-folders-for-dynamic-map-services

green3g committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee commented:
@roemhildtg 

>  I don't think I can personally utilize this in my app at the moment because of the way my groups are set up...

Do you think we should attempt to support both approaches to group layers? That adds to the complexity but certainly is possible.

> Which leads me to my second issue:
I'm not sure why at the moment but when I turn on the folder, like Documentation, it turns on all the other group layers, which renders about 30 layers. I wouldn't expect this to happen and it looks to me like a bug. I'll need to look into it a bit more. I can pm you with a link to the service to test on.

Did you find time to look into this?

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
fix how visiblity of sublayers is determined.
some code cleanup.

tmcgee commented:
@roemhildtg Using the test Map Service you provided to me (thanks), I have corrected how sub layer visibility is determined from their check boxes. Please give it another test and let me know how it works for you.

green3g committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

green3g commented:
> Do you think we should attempt to support both approaches to group layers? That adds to the complexity but certainly is possible.

I'm not a big fan of this, the dynamic layers have already become pretty complex and having two different cases makes it more difficult. I think the better solution would be to configure our layer setup to benefit from the new style. That being said though, I know there are others who do this same style of set up where they might have extra layers in a group that are invisible by default and might not have control over how the default visibility is published. 

> Please give it another test and let me know how it works for you.

It still looks like its having issues. Try this and let me know if you see the same thing:

Turn on the "Sanitary System" group layer check box. This will check everything under sanitary system. (which is expected). But what also happens is all of the other sub layers in other groups become visible on the map also. 

I'll starting digging in deeper later today to figure out why this is.

green3g commented:
Ok, so here's the question. Take a layer that has a parent layer invisible by default. Do we want the sub layer visibility to be not checked by default if its parent group is off by default? (easier)

Or do we want it to be checked, but excluded from the visible layers array when another layer's visibility changes? (harder)

tmcgee commented:
Maybe there's an issue with the group code since I am not as familiar with the data. But I think that part is working and there is a different issue that still needs to be resolved - one that has been [around for a while](https://github.com/cmv/cmv-app/issues/605) when groups are included in `ImageParameters` for a layer.

Here's what I *think* I see: The map is being presented initially based on the group layer visibility set in your service. The layerControl and the map are out of sync at this point (same as with ImageParameters problem). Once you change the visibility of any layer in layerControl (or from elsewhere?), the correct layers are made visible and represented accurately in the layerControl (same as with ImageParameters problem). Please confirm that is what you see as well. Then we can discuss a solution.

green3g commented:
> The map is being presented initially based on the group layer visibility set in your service. The layerControl and the map are out of sync at this point (same as with ImageParameters problem). 

Yes, that's what I'm seeing

tmcgee commented:
Alternative to your 2 options: Refresh the dynamic layer once it is loaded in layerControl. This "fixes" the visibility based on what is displayed in layerControl. Obvious downside is a new request to the server.

None of these options fix the ImageParameters "problem" if the layerControl widget is not included in the app. But is that a separate problem that needs code to "fix" or is it just a configuration challenge?

tmcgee commented:
FWIW: I do think this approach (tri-state check boxes for group layers) is a solution worth pursuing as it is more familiar to non-GIS users. The UI is quite common in other applications unlike the ArcGIS approach which is not at all common. Hopefully you agree and hopefully we can minimize any pain that might be inflicted along the way. :smile:

green3g commented:
>This "fixes" the visibility based on what is displayed in layerControl

This has been mentioned before, "fixing" the api's poor implementation. 

I think it would have negative side effects though, like if other widgets are using the event handlers for visible layers they would get different results if layer control is included or excluded. 

grrr, why did they make this so difficult?

>  I do think this approach (tri-state check boxes for group layers) is a solution worth pursuing

I do too and agree with you!  Maybe my app is just too big... How would you suggest improving my situation with the new controls? Maybe this is a side topic, and a pm would be more appropriate ?

tmcgee commented:
P.S. I am not liking the css I choose for the check boxes. They look oversized and the css does not work cleanly with other themes (like using calcite instead of flat). But that's a conversation for another day and another repo (cmv-themes?)

tmcgee commented:
how about a `pre-load` or `post-load` method/event for dynamic Map Services that strips the group ids out of the `visiblelayers` array before or just after it is added to the map? I think that would address all scenarios and the app would be more consistent across the board. The app however would not be consistent with what is published in the service with respect to group layer `visibility` which is okay by me. :wink:

Maybe we should set this aspect aside as a known issue and address separately since it isn't specific to layerControl?

tmcgee commented:
Could be combined with a Mixin that has been discussed before that detects the availability of services and removes them from the operational layers when they are not accessible. An [oldie](https://github.com/cmv/cmv-app/issues/339) but goodie.

tmcgee commented:
>  I think that would address all scenarios and the app would be more consistent across the board. 

I was wrong in this statement. Consider the case of dragging a dynamic Map Service url onto the map with the drag-and-drop widget. I have an approach in mind specific for the layerControl widget when it knows about the layer being added to the map (initially or by d-n-d) but not other scenarios. Will look into that.

More thought is needed on handling other scenarios where layerControl is not aware of the layer.

tmcgee commented:
Of course the first thing I considered won't work when the layerControl isn't loaded initially `preload: false`. At least that aspect is configurable so the app developer can preload the widget when he/she knows that there may be "special" dynamic Map Services with group layers.

green3g committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Remove sublayer ids for groups in dynamic layers after the layer is  loaded.

tmcgee committed:
remove console.log statement

tmcgee commented:
@roemhildtg I created a new method in _MapMixin to remove the group layers when the layer is loaded. How does this look now?

green3g commented:
Okay, I just had some time to test quickly, but for whatever reason, I can't get any layers to display using that test layer I shared with you. The first request appears to request all of the group layers (which requests every layer in the map service) and subsequent requests only request -1 (no layers). 

Edit: nope sorry that was an incorrect conclusion. I think it might be working as expected at the moment.

tmcgee commented:
> it might be working as expected at  the moment

live in the moment! :wink:

green3g commented:
Okay, tested again, and here's what I've found.

 - Default visibility is respected on initial load :+1: 
 - Once a layer checkbox is toggled, default visibility in groups is not respected anymore. If sublayers are published as visible in other groups, but the parent group layer is not visible by default, these other group's sublayers will become visible (this is the breaking change I do believe) :fried_egg: 

I'm okay with this behavior, but users will need to know that if they are using groups to control sublayer visibility, they should republish their services **without sublayers turned on**. This way, when a user of the app turns on a group, it will turn on all sublayers without turning on other sublayers. 

Does this make sense? Should we merge as is :question: 

----

Side note: I've also discovered that the 4.x api doesn't handle this correctly either. In the 4.x case, any sublayer set as visible when published will appear in the map regardless of parent group layer state.

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

duckblaster commented:
I would not use 3 state checkboxes, and have a layer be visible if it and it's parent layers are checked. If it is checked but any of it's parents aren't then it should not be visible. This would make things behave like layers in Photoshop and other graphics programs (GIMP, Inkscape, every other graphics program I have ever seen, many other programs with layers).

tmcgee commented:
@duckblaster thanks. your input is appreciated.

This update stems from multiple client requests to improve the clarity of the UI - especially when there are many nested/grouped sub-layers which is common in their environment. The most frequent request I have heard is "make the layer list work like Google Earth". For better or for worse, Google Earth is used widely and folks are familiar with it. And for worse, many call it a "GIS". 😢 

![image](https://user-images.githubusercontent.com/200780/29879590-52fea5ba-8d5a-11e7-9b04-d1da78e41e7d.png)

Tri-state checkboxes are a common UI element in both Windows and OS-X Operating systems and documented good practice in their UI/UX guidelines. So my thinking is it should not feel foreign to the majority of users. My first few customers that are using tri-state checkboxes in the layerControl and a few other widgets have responded favorably. 

Having said all that, I started this change knowing it would not meet everyone's needs and design objectives. The "C" in CMV is the most important aspect to me so the end-goal is to make this more configurable. With only a few code changes and a `tristate-checkbox: true/false` option, the UI of the LayerControl widget could be displayed in the way you and your users prefer.  I encourage you to contribute to this PR or a subsequent PR to help make that happen..

duckblaster committed:
Add option to use alternative visibility mode instead of triStateTree

duckblaster committed:
Fix lint errors

duckblaster commented:
I have sent a pull request to add that option, any improvements are appreciated especially the while loop in Dynamic.js (I think the `oldLayer !== currentLayer` check can go, but I left it in as insurance since I locked up the browser a few times)

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge pull request #766 from duckblaster/feature/new-group-folders-for-dynamic-map-services

Add option to use alternative visibility mode instead of triStateTree for dynamic layer group folders

tmcgee commented:
Discussion will continue here after merging from duckblaster's repo.

There is an issue with Identify widget related to the new group layer visibility approach. I will be addressing that in this branch.

tmcgee committed:
remove unused code and properties.

green3g committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

green3g committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
change property name to something more universal

The property can be unrelated to just the checkboxes in the layerControl widget so change it to something perhaps more clear.

tmcgee committed:
Simplify _setFolderCheckbox method

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Extend support for ignoring group layer visibility to Identify widget

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee commented:
@roemhildtg This is due for another review and discussion. I merged in the functionality from @duckblaster's branch many months ago before falling into abyss of the fire and parental care.

green3g committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge remote-tracking branch 'origin/feature/new-group-folders-for-dynamic-map-services' into feature/new-group-folders-for-dynamic-map-services

green3g commented:
Yeah, it appears to be good to go with several tests I was able to find online. I haven't revisited this in some time with the cities' layers but I remember the edge case had issues before where if I have a group layer turned off with sublayers turned on. I'll revisit and follow up.

green3g commented:
So...yes. That definitely still happens. I know you don't have a great way to test it though...every example I could find basically has everything visible by default. 

I also don't know how common of a practice it is with cmv's users to group sublayers in a "not checked" parent layer.

tr3vorm commented:
We often use the scenario where sublayers are enabled, with parent layer off by default. It just means that when a user does enable the parent, the pertinent underlying layers are shown. Just adding my +1 for this desired behaviour.

tmcgee commented:
@roemhildtg @tr3vorm 

It has been awhile since I have thought about this too. If I recall correctly, I addressed this by adding the [`ignoreDynamicGroupVisibility](https://githu'b.com/cmv/cmv-app/pull/747/files#diff-17e366da090c9f7fd8aadfd28d9465ddR162) as a toggle. The intent of 
``` javascript
ignoreDynamicGroupVisibility: false
```
is to restore the original Esri group layer behavior as you describe. This was merged from @duckblaster's branch

 Let me know if this works the way you would expect.

green3g commented:
Oh nice! Yeah that does work 👍

tmcgee commented:
@roemhildtg 👍 

Anything remaining to do before this is pushed out within a 2.0 beta 3 release?

duckblaster commented:
I noticed that you `removedsetAllVisibleLayers` in [in this commit](https://github.com/cmv/cmv-app/pull/747/commits/62d73116b837240b03f6ebb009547e14d41ecdab), I believe this was something I added to make @roemhildtg's AppSettings widget save sublayer visibility correctly, by pointing it at `layer.allVisibleLayers` instead of `layer.visibleLayers`

tmcgee commented:
@duckblaster Is that something you think should be added to CMV core for other purposes or is it specific to that widget only?

duckblaster commented:
It would be helpful to have the layer visibility as sent to esri, and as set by the user, and load them as appropriate

duckblaster commented:
It would need to be in core CMV, I don't know how you would get the info otherwise

tmcgee commented:
ok, though I suggest perhaps we consider it be done differently (and in a different PR?). I'm not a fan of functionality in one widget (AppSettings) dependent on functionality created by another widget (LayerControl).

duckblaster commented:
The other part of the code used for this is
```
            topic.publish('layerControl/setAllVisibleLayers', {
                id: layer.id,
                allVisibleLayers: allVisibleLayers
            });
```
The AppSettings widget probably should be talking directly to the LayerControl widget, rather than going via the layer, but I needed to just get something that worked (and I wan't that clear on exactly how everything hooked together)

duckblaster commented:
My [changes](https://github.com/duckblaster/cmv-widgets/commit/4a9987e31b2b89a4bcbbdf7a86d1f4b2c0f7cb47) to the AppSettings widget (for reference)

tmcgee commented:
@roemhildtg what are your thoughts on the connection between AppSettings and LayerControl? My feeling is whatever changes may come from that specific discussion, it should be handled in a separate PR as it is not directly related to the intent of this PR.  thoughts?

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

tmcgee committed:
Merge branch 'develop' into feature/new-group-folders-for-dynamic-map-services

green3g commented:
Sorry, forgot to get back to you on this...

>  it should be handled in a separate PR as it is not directly related to the intent of this PR. thoughts?

Yes, I agree. I'm not a super big fan of my app setting implementation anyways, there's still some race conditions in there somewhere where the settings aren't loaded correctly at times.

tmcgee commented:
@roemhildtg I had good results moving Brian Bunker's [MapNavigationHash widget](https://github.com/BrianBunker/cmv-widgets/tree/master//MapNavigationHash) to a mixin to address race conditions with that widget. Perhaps that approach would work for your Settings widget too.

tmcgee commented:
@roemhildtg @duckblaster @tr3vorm Any outstanding items/concerns in this PR? Would love to get it committed.

duckblaster commented:
Looks good to me

green3g commented:
Yup, lets merge this.

tmcgee commented:
@roemhildtg @duckblaster thanks!

@roemhildtg please merge. 😄
